hashing/src/indexed_merkle_proof.rs

```diff
@@ -0,0 +1,465 @@
+        let raw_root = if let Some(leaf_hash) = hashes.next().cloned() {
+            // Compute whether to hash left or right for the elements of the Merkle proof.
+            // This gives a path to the value with the specified index.
+            // We represent this path as a sequence of 64 bits. 1 here means "hash right".
+            let mut path: u64 = 0;
+            let mut n = self.count;
+            let mut i = self.index;
+            while n > 1 {
+                path <<= 1;
+                let pivot = 1u64 << (63 - (n - 1).leading_zeros());
+                if i < pivot {
+                    n = pivot;
+                } else {
+                    path |= 1;
+                    n -= pivot;
+                    i -= pivot;
+                }
+            }
+
+            // Compute the raw Merkle root by hashing the proof from leaf hash up.
+            hashes.fold(leaf_hash, |acc, hash| {
+                let digest = if (path & 1) == 1 {
+                    Digest::hash_pair(hash, &acc)
+                } else {
+                    Digest::hash_pair(&acc, hash)
+                };
+                path >>= 1;
+                digest
+            })
+        } else {
+            Digest::SENTINEL_MERKLE_TREE
+        };
+
+        // The Merkle root is the hash of the count with the raw root.
+        Digest::hash_pair(count.to_le_bytes(), raw_root)
+    }
+
+    pub(crate) fn merkle_proof(&self) -> &[Digest] {
+        &self.merkle_proof
+    }
+
+    // Proof lengths are never bigger than 65 is because we are using 64 bit counts
+    fn compute_expected_proof_length(&self) -> u8 {
+        if self.count == 0 {
+            return 0;
+        }
+        let mut l = 1;
+        let mut n = self.count;
+        let mut i = self.index;
+        while n > 1 {
+            let pivot = 1u64 << (63 - (n - 1).leading_zeros());
+            if i < pivot {
+                n = pivot;
+            } else {
+                n -= pivot;
+                i -= pivot;
+            }
+            l += 1;
+        }
+        l
+    }
+
+    pub(crate) fn verify(&self) -> Result<(), MerkleVerificationError> {
+        if self.index >= self.count {
+            return Err(MerkleVerificationError::IndexOutOfBounds {
+                count: self.count,
+                index: self.index,
+            });
+        }
+        let expected_proof_length = self.compute_expected_proof_length();
+        if self.merkle_proof.len() != expected_proof_length as usize {
+            return Err(MerkleVerificationError::UnexpectedProofLength {
+                count: self.count,
+                index: self.index,
+                expected_proof_length,
+                actual_proof_length: self.merkle_proof.len(),
+            });
+        }
+        Ok(())
+    }
+
+    #[cfg(test)]
+    pub(crate) fn inject_merkle_proof(&mut self, merkle_proof: Vec<Digest>) {
+        self.merkle_proof = merkle_proof;
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
```suggestion
mod tests {
```

Reply:
Fixed (also in other places) in https://github.com/casper-network/casper-node/pull/2119/commits/05292085011a457a8fbbe8016e513023b8b9f509.